File: Lecture_2_Arrays/problem_set/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

#define FORMULA(L, S) 0.0588 * L - 0.296 * S - 15.8
// L is the average number of letters per 100 words in the text:
// that is, the number of letters divided by the number of words, all multiplied by 100.
// S is the average number of sentences per 100 words in the text:
// that is, the number of sentences divided by the number of words, all multiplied by 100.

int get_grade(string text);

int main(void)
{
    // 获取用户输入
    string text = get_string("Text: ");

    // 计算文本等级
    int grade = get_grade(text);

    // 输出等级
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

int get_grade(string text)
{
    int letters = 0;
    int sentences = 0;
    int words = 0;
    int len = strlen(text);
    for (int i = 0; i < len; i++)
    {
        if (text[i] == '.' || text[i] == '?' || text[i] == '!')
        {
            sentences++;
        }
        if (isalpha(text[i]))
        {
            letters++;
        }
        if (text[i] == ' ' || i == len - 1)
        {
            words++;
        }
    }
    float L = ((float) letters / words) * 100;
    float S = ((float) sentences / words) * 100;

    return round(FORMULA(L, S)); // 四舍五入
}

File: Lecture_6_Python/problem_set/sentimental-readability/readability.py
from cs50 import get_string
import re


def get_grade(text):
    # re.split() 是 Python 中 re 模块（正则表达式模块）提供的一个函数
    # 它用于根据匹配正则表达式的模式来分割字符串。这个函数会返回一个列表，其中包含了被分割的部分
    sentences = len(re.split('[.!?]', text)) - 1
    # 减1是因为分割后的最后一个是空字符串
    # 原始字符串以"."结尾，所以最后一个分隔符后面没有更多的字符
    # 如果你不想在结果中包含空字符串，可以使用列表推导式来过滤它们
    # tokens = [token for token in re.split(r'[.?!]', text) if token]
    words = len(re.split(' ', text))
    letters = 0
    for c in text:
        if c.isalpha():
            letters += 1
    L = letters / words * 100
    S = sentences / words * 100
    return round(0.0588 * L - 0.296 * S - 15.8)


text = get_string("Text: ")
grade = get_grade(text)

if grade < 1:
    print("Before Grade 1")
elif grade >= 16:
    print("Grade 16+")
else:
    print(f"Grade {grade}")
